node/core/approval-voting/src/approval_checking.rs

```diff
@@ -938,6 +958,145 @@ mod tests {
+            // Single validator with last-minute approval.
+            NoShowTest {
+                assignments: vec![(ValidatorIndex(1), 20)],
+                approvals: vec![1],
+                clock_drift: 10,
+                no_show_duration: 10,
+                drifted_tick_now: 20,
+                exp_no_shows: 0,
+                exp_next_no_show: None,
+            },
+            // Single validator with late approval.
+            NoShowTest {
+                assignments: vec![(ValidatorIndex(1), 21)],
+                approvals: vec![1],
+                clock_drift: 10,
+                no_show_duration: 10,
+                drifted_tick_now: 20,
+                exp_no_shows: 0,
+                exp_next_no_show: None,
+            },
+            // Two almost late validators, next_no_show ordered last.
+            NoShowTest {
+                assignments: vec![(ValidatorIndex(1), 22), (ValidatorIndex(2), 21)],
+                approvals: vec![],
+                clock_drift: 10,
+                no_show_duration: 10,
+                drifted_tick_now: 20,
+                exp_no_shows: 0,
+                exp_next_no_show: Some(31),
+            },
+            // Two almost late validators, next_no_show ordered first.
+            NoShowTest {
+                assignments: vec![(ValidatorIndex(1), 21), (ValidatorIndex(2), 22)],
+                approvals: vec![],
+                clock_drift: 10,
+                no_show_duration: 10,
+                drifted_tick_now: 20,
+                exp_no_shows: 0,
+                exp_next_no_show: Some(31),
+            },
+            // One almost-late validator, one no_show validator, and one
+            // approving validator.
+            NoShowTest {
+                assignments: vec![(ValidatorIndex(1), 21), (ValidatorIndex(2), 20), (ValidatorIndex(3), 20)],
+                approvals: vec![3],
+                clock_drift: 10,
+                no_show_duration: 10,
+                drifted_tick_now: 20,
+                exp_no_shows: 1,
+                exp_next_no_show: Some(31),
+            },
+            // Three no_show validators.
+            NoShowTest {
+                assignments: vec![(ValidatorIndex(1), 20), (ValidatorIndex(2), 20), (ValidatorIndex(3), 20)],
+                approvals: vec![],
+                clock_drift: 10,
+                no_show_duration: 10,
+                drifted_tick_now: 20,
+                exp_no_shows: 3,
+                exp_next_no_show: None,
+            },
+            // Three approving validators.
+            NoShowTest {
+                assignments: vec![(ValidatorIndex(1), 20), (ValidatorIndex(2), 20), (ValidatorIndex(3), 20)],
+                approvals: vec![1, 2, 3],
+                clock_drift: 10,
+                no_show_duration: 10,
+                drifted_tick_now: 20,
+                exp_no_shows: 0,
+                exp_next_no_show: None,
+            },
+        ];
+
+        for test in tests {
```

Comment:
style nit: I'd prefer each test to get its own `#[test]` function which uses a helper function as the body of this loop

Reply:
Good to know, will make not of that going forward.

Reply:
@rphmeier force pushed the change to create each test case individually. Should be good to go.